File: pfm-sdk/src/main/java/com/tink/pfmsdk/view/ParallaxHeaderScrollListener.java
package com.tink.pfmsdk.view;

import android.view.View;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.RecyclerView.OnScrollListener;
import java.util.List;

public class ParallaxHeaderScrollListener extends OnScrollListener {

    private List<View> headers;
    private float headerHeight;
    private int scrolledY;

    public ParallaxHeaderScrollListener(List<View> headers, float headerHeight) {
        this.headers = headers;

        this.headerHeight = headerHeight;
    }

    @Override
    public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
        super.onScrolled(recyclerView, dx, dy);

        if (dy == 0) { // Layout event happened, recalculate
            scrolledY = recyclerView.computeVerticalScrollOffset();
        }

        scrolledY += dy;
        float alpha = ((headerHeight - scrolledY * 1.66f) / (headerHeight));
        int topMargin = -scrolledY / 8;

        for (View header : headers) {
            header.setTranslationY(topMargin);
            header.setAlpha(alpha);
        }
    }
}

File: pfm-ui/src/main/java/com/tink/pfmui/view/ParallaxHeaderScrollListener.kt
package com.tink.pfmui.view

import android.view.View
import androidx.recyclerview.widget.RecyclerView

internal class ParallaxHeaderScrollListener(
    private val headers: List<View>,
    private val headerHeight: Float
) : RecyclerView.OnScrollListener() {

    private var scrolledY = 0

    override fun onScrolled(
        recyclerView: RecyclerView,
        dx: Int,
        dy: Int
    ) {
        super.onScrolled(recyclerView, dx, dy)
        if (dy == 0) {
            // Layout event happened, recalculate
            scrolledY = recyclerView.computeVerticalScrollOffset()
        }
        scrolledY += dy
        val alpha = (headerHeight - scrolledY * 1.66f) / headerHeight
        val topMargin = -scrolledY / 8
        for (header in headers) {
            header.translationY = topMargin.toFloat()
            header.alpha = alpha
        }
    }
}